### ISSUE TITLE
Bad error messages when shadowing variables

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#163]
This code

```
val swap_add_sub : #int:Type -> r1:(ref int) -> r2:(ref int) -> St unit
      (requires (fun h -> contains h r1 /\ contains h r2))
      (ensures (fun h' _ h -> True))
let swap_add_sub r1 r2 =
  write r1 (read r1 + read r2);
  write r2 (read r1 - read r2);
  write r1 (read r1 - read r2)
```

Causes this error message:

```
swap.fst(18,12-18,21): Subtyping check failed; expected type int; got type int
```

**Comment by catalin-hritcu** [comment]
My bad, I was rebinding int after a copy place + replace. Thanks to Nick for spotting this yesterday.

Still, our error messages can be very confusing when there is shadowing involved (this is just an extreme example). Would it be an option to detect such problems in error messages and then add the declaration point to each shadowed variable?

**Comment by catalin-hritcu** [comment]
Thanks to recent improvements by @protz the error messages are better now:
```
Subtyping check failed; expected type int#3835; got type Prims.int (see /home/hritcu/Projects/fstar/pub/examples/bug-reports/bug163.fst(0,0-0,0))
```
Closing.

**Comment by aseemr** [comment]
There are other places in the type checker where this fix needs to go, for example, this is something I encountered today:

```
module C

open FStar.UInt16

assume val foo: unit -> FStar.UInt8.t

let bar () :t =
  let x = foo () in
  x
```
```
$ fstar.exe C.fst
.\C.fst(8,6-8,7) : (Error) Expected expression of type "t"; got expression "x" of type "t"
```

Reopening.

**Comment by aseemr** [comment]
Hopefully fixed with 91e509806e4f41c6b12cd3c7446173e25aed36c6.
